File: app/src/main/java/com/porarrirr/sumahohikakuku/viewmodel/GenerateComparisonUseCase.kt
package com.porarrirr.sumahohikakuku.viewmodel

import com.porarrirr.sumahohikakuku.model.ComparisonResults
import com.porarrirr.sumahohikakuku.model.LensProcessingInput
import com.porarrirr.sumahohikakuku.model.SensorSpec
import com.porarrirr.sumahohikakuku.model.calculateNativeSensorMetrics
import com.porarrirr.sumahohikakuku.model.computeProcessedDevice
import kotlin.math.abs

internal data class ParsedOpticalLensInput(
    val endFocal: Double,
    val endFNumber: Double
)

internal fun parseOptionalOpticalLensInput(
    lens: LensInputState,
    nativeFocal: Double,
    fNumber: Double
): ParsedOpticalLensInput? {
    val endFocalText = lens.opticalEndFocalLength.trim()
    val endFNumberText = lens.endFNumber.trim()
    if (endFocalText.isBlank()) {
        return if (endFNumberText.isBlank()) {
            ParsedOpticalLensInput(endFocal = nativeFocal, endFNumber = fNumber)
        } else {
            null
        }
    }

    val endFocal = endFocalText.toDoubleOrNull()?.takeIf { it >= nativeFocal } ?: return null
    val endFNumber = if (endFNumberText.isBlank()) {
        fNumber
    } else {
        endFNumberText.toDoubleOrNull()?.takeIf { it > 0.0 } ?: return null
    }
    return ParsedOpticalLensInput(endFocal = endFocal, endFNumber = endFNumber)
}

data class GeneratedComparison(
    val results: ComparisonResults?,
    val focalLengths: List<Double>,
    val selectedFocalLength: Double
)

class GenerateComparisonUseCase {
    fun generate(
        devices: List<DeviceInputState>,
        availableSensors: List<SensorSpec>,
        selectedFocalLength: Double,
        defaultFocalLengths: List<Double>,
        fallbackDeviceName: (Int) -> String
    ): GeneratedComparison {
        val sensorLookup = availableSensors.associateBy { it.value }
        val nativeFocals = devices.flatMap { device ->
            device.lenses.flatMap lensFocals@ { lens ->
                val focal = lens.nativeFocalLength.toDoubleOrNull()?.takeIf { it > 0.0 }
                    ?: return@lensFocals emptyList()
                val fNumber = lens.fNumber.toDoubleOrNull()?.takeIf { it > 0.0 }
                    ?: return@lensFocals listOf(focal)
                val opticalEnd = parseOptionalOpticalLensInput(lens, focal, fNumber)
                    ?.endFocal
                    ?.takeIf { it != focal }
                listOfNotNull(
                    focal,
                    opticalEnd
                )
            }
        }
        val focalGrid = (defaultFocalLengths + nativeFocals).distinct().sorted()

        val processedDevices = devices.mapIndexedNotNull { index, device ->
            val sanitizedName = device.name.ifBlank { fallbackDeviceName(index + 1) }
            val rawLenses = device.lenses.mapNotNull { lens ->
                val focal = lens.nativeFocalLength.toDoubleOrNull()
                val fNumber = lens.fNumber.toDoubleOrNull()
                if (focal == null || focal <= 0.0 || fNumber == null || fNumber <= 0.0) return@mapNotNull null
                val sensorSpec = sensorLookup[lens.selectedSensorValue]
                val manualDescriptor = if (lens.usesManualSensor) lens.manualSensorDescriptor else null
                val metrics = calculateNativeSensorMetrics(sensorSpec, manualDescriptor)
                if (metrics.areaSqMm <= 0.0 || metrics.diagonalMm <= 0.0) return@mapNotNull null
                val optical = parseOptionalOpticalLensInput(lens, focal, fNumber)
                    ?: return@mapNotNull null
                LensProcessingInput(
                    nativeFocalLength35mm = focal,
                    fNumber = fNumber,
                    sensorMetrics = metrics,
                    opticalEndFocalLength35mm = optical.endFocal,
                    endFNumber = optical.endFNumber
                )
            }
            computeProcessedDevice(sanitizedName, device.colorHex, rawLenses, focalGrid)
        }

        if (processedDevices.isEmpty()) {
            return GeneratedComparison(
                results = null,
                focalLengths = defaultFocalLengths,
                selectedFocalLength = defaultFocalLengths.first()
            )
        }

        val availableFocalLengths = processedDevices
            .flatMap { device -> device.metricsByFocalLength.map { it.focalLength35mm } }
            .distinct()
            .sorted()

        if (availableFocalLengths.isEmpty()) {
            return GeneratedComparison(
                results = null,
                focalLengths = defaultFocalLengths,
                selectedFocalLength = defaultFocalLengths.first()
            )
        }

        val nearestFocal = availableFocalLengths.minByOrNull { abs(it - selectedFocalLength) }
            ?: availableFocalLengths.first()

        return GeneratedComparison(
            results = ComparisonResults(
                focalLengths = availableFocalLengths,
                devices = processedDevices
            ),
            focalLengths = availableFocalLengths,
            selectedFocalLength = nearestFocal
        )
    }
}

File: ios/src/Domain/GenerateComparisonUseCase.swift
import Foundation

public struct GeneratedComparison: Equatable {
    public let results: ComparisonResults?
    public let focalLengths: [Double]
    public let selectedFocalLength: Double
}

struct ParsedOpticalLensInput {
    let endFocal: Double
    let endFNumber: Double
}

func parseOptionalOpticalLensInput(
    lens: LensInputState,
    nativeFocal: Double,
    fNumber: Double
) -> ParsedOpticalLensInput? {
    let endFocalText = lens.opticalEndFocalLength.trimmingCharacters(in: .whitespacesAndNewlines)
    let endFNumberText = lens.endFNumber.trimmingCharacters(in: .whitespacesAndNewlines)
    if endFocalText.isEmpty {
        return endFNumberText.isEmpty
            ? ParsedOpticalLensInput(endFocal: nativeFocal, endFNumber: fNumber)
            : nil
    }

    guard let endFocal = Double(endFocalText), endFocal >= nativeFocal else {
        return nil
    }
    let endFNumber: Double
    if endFNumberText.isEmpty {
        endFNumber = fNumber
    } else {
        guard let parsedEndFNumber = Double(endFNumberText), parsedEndFNumber > 0.0 else {
            return nil
        }
        endFNumber = parsedEndFNumber
    }
    return ParsedOpticalLensInput(endFocal: endFocal, endFNumber: endFNumber)
}

public class GenerateComparisonUseCase {
    public init() {}
    
    public func generate(
        devices: [DeviceInputState],
        availableSensors: [SensorSpec],
        selectedFocalLength: Double,
        defaultFocalLengths: [Double],
        fallbackDeviceName: (Int) -> String
    ) -> GeneratedComparison {
        var sensorLookup: [String: SensorSpec] = [:]
        for sensor in availableSensors {
            sensorLookup[sensor.value] = sensor
        }
        
        let nativeFocals = devices.flatMap { device in
            device.lenses.flatMap { lens -> [Double] in
                guard let focal = Double(lens.nativeFocalLength), focal > 0.0 else {
                    return []
                }
                guard let fNumber = Double(lens.fNumber), fNumber > 0.0 else {
                    return [focal]
                }
                let opticalEnd = parseOptionalOpticalLensInput(
                    lens: lens,
                    nativeFocal: focal,
                    fNumber: fNumber
                )?.endFocal
                return [
                    focal,
                    opticalEnd.flatMap { $0 == focal ? nil : $0 }
                ].compactMap { $0 }
            }
        }
        
        var focalGridSet = Set<Double>()
        for f in defaultFocalLengths {
            focalGridSet.insert(f)
        }
        for f in nativeFocals {
            focalGridSet.insert(f)
        }
        let focalGrid = focalGridSet.sorted()
        
        var processedDevices: [ProcessedDevice] = []
        for (index, device) in devices.enumerated() {
            let name = device.name.trimmingCharacters(in: .whitespacesAndNewlines)
            let sanitizedName = name.isEmpty ? fallbackDeviceName(index + 1) : name
            
            var rawLenses: [LensProcessingInput] = []
            for lens in device.lenses {
                guard let focal = Double(lens.nativeFocalLength), focal > 0.0,
                      let fNumber = Double(lens.fNumber), fNumber > 0.0 else {
                    continue
                }
                let sensorSpec = sensorLookup[lens.selectedSensorValue]
                let manualDescriptor = lens.usesManualSensor ? lens.manualSensorDescriptor : nil
                let metrics = calculateNativeSensorMetrics(sensorSpec: sensorSpec, manualDescriptor: manualDescriptor)
                if metrics.areaSqMm <= 0.0 || metrics.diagonalMm <= 0.0 {
                    continue
                }
                guard let optical = parseOptionalOpticalLensInput(lens: lens, nativeFocal: focal, fNumber: fNumber) else {
                    continue
                }
                rawLenses.append(LensProcessingInput(
                    nativeFocalLength35mm: focal,
                    fNumber: fNumber,
                    sensorMetrics: metrics,
                    opticalEndFocalLength35mm: optical.endFocal,
                    endFNumber: optical.endFNumber
                ))
            }
            
            if let processed = computeProcessedDevice(name: sanitizedName, colorHex: device.colorHex, rawLenses: rawLenses, focalLengths: focalGrid) {
                processedDevices.append(processed)
            }
        }
        
        if processedDevices.isEmpty {
            return GeneratedComparison(
                results: nil,
                focalLengths: defaultFocalLengths,
                selectedFocalLength: defaultFocalLengths.first ?? 14.0
            )
        }
        
        var availableFocalLengthsSet = Set<Double>()
        for device in processedDevices {
            for metric in device.metricsByFocalLength {
                availableFocalLengthsSet.insert(metric.focalLength35mm)
            }
        }
        let availableFocalLengths = availableFocalLengthsSet.sorted()
        
        if availableFocalLengths.isEmpty {
            return GeneratedComparison(
                results: nil,
                focalLengths: defaultFocalLengths,
                selectedFocalLength: defaultFocalLengths.first ?? 14.0
            )
        }
        
        var nearestFocal = availableFocalLengths.first!
        var minDiff = abs(nearestFocal - selectedFocalLength)
        for focal in availableFocalLengths {
            let diff = abs(focal - selectedFocalLength)
            if diff < minDiff {
                minDiff = diff
                nearestFocal = focal
            }
        }
        
        return GeneratedComparison(
            results: ComparisonResults(
                focalLengths: availableFocalLengths,
                devices: processedDevices
            ),
            focalLengths: availableFocalLengths,
            selectedFocalLength: nearestFocal
        )
    }
}
